Mathlib/Analysis/BoxIntegral/UnitPartition.lean

```diff
@@ -0,0 +1,286 @@
+    x ∈ box n ν ↔ ∀ i, ν i < n * x i ∧ n * x i ≤ ν i + 1 := by
+  have h_npos : 0 < (n:ℝ) := Nat.cast_pos.mpr <| PNat.pos n
+  simp_rw [mem_box_iff, ← _root_.le_div_iff₀' h_npos, ← div_lt_iff₀' h_npos, add_div]
+
+/-- The tag of a `unitPartition.Box`. -/
+abbrev tag (ν : ι → ℤ) : ι → ℝ := fun i ↦ (ν i + 1) / n
+
+@[simp]
+theorem tag_apply (ν : ι → ℤ) (i : ι) : tag n ν i = (ν i + 1) / n := rfl
+
+theorem tag_injective : Function.Injective (fun ν : ι → ℤ ↦ tag n ν) := by
+  refine fun _ _ h ↦ funext_iff.mpr fun i ↦ ?_
+  have := congr_arg (fun x ↦ x i) h
+  field_simp at this
+  exact this
+
+theorem tag_mem (ν : ι → ℤ) :
+    tag n ν ∈ box n ν := by
+  refine mem_box_iff.mpr fun _ ↦ ?_
+  rw [tag, add_div]
+  exact ⟨by norm_num, le_rfl⟩
+
+/-- For `x : ι → ℝ`, its index is the index of the unique `unitPartition.Box` to which
+it belongs. -/
+def index (x : ι → ℝ) (i : ι) : ℤ := ⌈n * x i⌉ - 1
+
+@[simp]
+theorem index_apply {x : ι → ℝ} (i : ι) :
+    index n x i = ⌈n * x i⌉ - 1 := rfl
+
+variable {n} in
+theorem mem_box_iff_index {x : ι → ℝ} {ν : ι → ℤ} :
+    x ∈ box n ν ↔ index n x = ν := by
+  simp_rw [mem_box_iff', funext_iff, index_apply, sub_eq_iff_eq_add, Int.ceil_eq_iff,
+    Int.cast_add, Int.cast_one, add_sub_cancel_right]
+
+@[simp]
+theorem index_tag (ν : ι → ℤ) :
+    index n (tag n ν) = ν := mem_box_iff_index.mp (tag_mem n ν)
+
+variable {n} in
+theorem disjoint {ν ν' : ι → ℤ} :
+    ν ≠ ν' ↔ Disjoint (box n ν).toSet (box n ν').toSet := by
+  rw [not_iff_not.symm, ne_eq, not_not, Set.not_disjoint_iff]
+  refine ⟨fun h ↦ ?_, fun ⟨x, hx, hx'⟩ ↦ ?_⟩
+  · exact ⟨tag n ν, tag_mem n ν, h ▸ tag_mem n ν⟩
+  · rw [← mem_box_iff_index.mp hx, ← mem_box_iff_index.mp hx']
+
+theorem box_injective : Function.Injective (fun ν : ι → ℤ ↦ box n ν) := by
+  intro _ _ h
+  contrapose! h
+  exact Box.ne_of_disjoint_coe (disjoint.mp h)
+
+variable [Fintype ι]
+
+theorem diam_boxIcc (ν : ι → ℤ) :
+    Metric.diam (Box.Icc (box n ν)) ≤ 1 / n := by
+  refine ENNReal.toReal_le_of_le_ofReal (by positivity) ?_
+  rw [BoxIntegral.Box.Icc_eq_pi]
+  refine EMetric.diam_pi_le_of_le (fun i ↦ ?_)
+  rw [Real.ediam_Icc, box, add_sub_cancel_left, ENNReal.ofReal_div_of_pos (Nat.cast_pos.mpr n.pos),
+    ENNReal.ofReal_one]
+
+@[simp]
+theorem volume_box (ν : ι → ℤ) :
+    volume (box n ν : Set (ι → ℝ)) = 1 / n ^ card ι := by
+  simp_rw [volume_pi, BoxIntegral.Box.coe_eq_pi, Measure.pi_pi, Real.volume_Ioc, box,
+    add_sub_cancel_left, Finset.prod_const, ENNReal.ofReal_div_of_pos (Nat.cast_pos.mpr n.pos),
+    ENNReal.ofReal_one, ENNReal.ofReal_natCast, Finset.card_univ, one_div, ENNReal.inv_pow]
+
+theorem setFinite_index {s : Set (ι → ℝ)} (hs₁ : NullMeasurableSet s) (hs₂ : volume s ≠ ⊤) :
+    Set.Finite {ν : ι → ℤ | ↑(box n ν) ⊆ s} := by
+  refine (Measure.finite_const_le_meas_of_disjoint_iUnion₀ volume (ε := 1 / n ^ card ι)
+    (by norm_num) (As := fun ν : ι → ℤ ↦ (box n ν) ∩ s) (fun ν ↦ ?_) (fun _ _ h ↦ ?_) ?_).subset
```

Comment:
`by norm_num` should be `by positivity`, but `positivity` cannot deal with `ENNReal`s...